src/EnergyPlus/PhaseChangeModeling/HysteresisModel.cc

```diff
@@ -0,0 +1,344 @@
+	}
+
+	Real64 HysteresisPhaseChange::specHeat( Real64 temperaturePrev, Real64 temperatureCurrent, Real64 criticalTemperature, Real64 tau1, Real64 tau2, Real64 EnthalpyOld, Real64 EnthalpyNew ) {
+
+		//	Tc                  ! Critical (Melting/Freezing) Temperature of PCM
+		//	Tau1                ! Width of Melting Zone low
+		//	Tau2                ! Width of Melting Zone high
+		//	EnthalpyOld         ! Previos Timestep Nodal Enthalpy
+		//	EnthalpyNew         ! Current Timestep Nodal Enthalpy
+
+        Real64 Cp1 = this->specificHeatSolid;
+        Real64 Cp2 = this->specificHeatLiquid;
+        Real64 T = temperatureCurrent;
+
+        Real64 DEta1 = - ( this->totalLatentHeat * ( T - criticalTemperature ) * exp( -2 * abs( T - criticalTemperature ) / tau1 ) ) / ( tau1 * abs( T - criticalTemperature ) );
+        Real64 DEta2 = ( this->totalLatentHeat * ( T - criticalTemperature ) * exp( -2 * abs( T - criticalTemperature ) / tau2 ) ) / ( tau2 * abs( T - criticalTemperature ) );
+
+        if ( T < criticalTemperature ) {
+            return (Cp1 + DEta1);
+        } else if ( T == criticalTemperature ) {
+            return (EnthalpyNew - EnthalpyOld) / (temperatureCurrent - temperaturePrev);
+        } else if ( T > criticalTemperature ) {
+            return Cp2 + DEta2;
+        } else {
+            return 0;
+        }
+
+	}
+
+	void readAllHysteresisModels() {
+
+		// convenience variables
+		DataIPShortCuts::cCurrentModuleObject = "MaterialProperty:PhaseChangeHysteresis";
+		numHysteresisModels = InputProcessor::GetNumObjectsFound( DataIPShortCuts::cCurrentModuleObject );
+
+		// loop over all hysteresis input instances, if zero, this will simply not do anything
+		for ( int hmNum = 1; hmNum <= numHysteresisModels; ++ hmNum ) {
+
+			// just a few vars to pass in and out to GetObjectItem
+			int ioStatus;
+			int numAlphas;
+			int numNumbers;
+
+			// get the input data and store it in the Shortcuts structures
+			InputProcessor::GetObjectItem(
+				DataIPShortCuts::cCurrentModuleObject,
+				hmNum,
+				DataIPShortCuts::cAlphaArgs,
+				numAlphas,
+				DataIPShortCuts::rNumericArgs,
+				numNumbers,
+				ioStatus,
+				DataIPShortCuts::lNumericFieldBlanks,
+				DataIPShortCuts::lAlphaFieldBlanks,
+				DataIPShortCuts::cAlphaFieldNames,
+				DataIPShortCuts::cNumericFieldNames
+			);
+
+			// the input processor validates the numeric inputs based on the IDD definition
+			// still validate the name to make sure there aren't any duplicates or blanks
+			// blanks are easy: fatal if blank
+			if ( DataIPShortCuts::lAlphaFieldBlanks[0] ) {
+				ShowFatalError( "Invalid input for " + DataIPShortCuts::cCurrentModuleObject + " object: Name cannot be blank" );
+			}
+
+			// we just need to loop over the existing vector elements to check for duplicates since we haven't add this one yet
+			for ( auto & existingHysteresisModel : hysteresisPhaseChangeModels ) {
+				if ( DataIPShortCuts::cAlphaArgs( 1 ) == existingHysteresisModel.name ) {
```

Comment:
Adding to then comparing to a `std::set`/`std::unordered_set` or something would be better here. That would change it from `O(n^2)` to `O(n log n)`/`O(n)` but i'm fine with this as is for now.